crates/sui-core/src/scoring_decision.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::math::median;
+use arc_swap::ArcSwap;
+use narwhal_types::ReputationScores;
+use std::collections::HashMap;
+use std::sync::Arc;
+use sui_types::base_types::AuthorityName;
+use sui_types::committee::Committee;
+use tracing::warn;
+
+/// Updates list of authorities that are deemed to have low reputation scores by consensus
+/// these may be lagging behind the network, byzantine, or not reliably participating for any reason.
+/// We want to ensure that the remaining set of validators once we exclude the low scoring authorities
+/// is including enough stake for a quorum, at the very least. It is also possible that no authorities
+/// are particularly low scoring, in which case this will result in storing an empty list.
+///
+/// Rather than using hardcoded cutoff score values, which will need consistent maintenance and create
+/// a tight coupling between this code and the scoring code, we detect low-value outliers using
+/// all the validator scores. The outlier detection method is by using the adjusted
+/// median absolute deviation see https://en.wikipedia.org/wiki/Median_absolute_deviation for more
+/// details. This calculates a the median of the data, then the absolute deviations from the median
+/// for each authority, or the difference between the median and the score value. We then take the
+/// median of those absolute deviations for each authority, which is called the median absolute deviation (MAD).
+/// Once we have that value, if any authority's absolute deviation / ( MAD / C) < -K then it is deemed
+/// to be a low-value outlier. The values of C and K can be tweaked to change the sensitivity to outliers.
+/// They were chosen based on trial and error to produce reasonable results for score values in the
+/// order of magnitude of 10s - 1000s. If you increase C and decrease K you will see more values
+/// being included as outliers. As the scores get higher in value, outlier sensitivity tends to
+/// decrease using this method.
+///
+/// If we find that we have rated enough validators as low scoring such that we no longer have
+/// quorum with the remaining validators, then we either need to update this method's parameters,
+/// our general approach to finding outliers, or our network is in a bad state. If we need to update
+/// this code, we let it detect this and disable itself for safety reasons. If we have a bad network
+/// state then in the interest of making debugging and investigation easier, disabling the scoring
+/// mechanism will likely be helpful.
+pub fn update_low_scoring_authorities(
+    low_scoring_authorities: &ArcSwap<HashMap<AuthorityName, u64>>,
+    committee: Committee,
+    reputation_scores: ReputationScores,
+) {
+    if !reputation_scores.final_of_schedule {
+        return;
+    }
+    let mut score_list = vec![];
```

Comment:
As we discussed offline, at least while we are testing this, I would suggest adding some logging with all the scores we receive here per node to have better visibility on the operations.